Mathlib/ModelTheory/Algebra/Field/IsAlgClosed.lean

```diff
@@ -0,0 +1,285 @@
+/-
+Copyright (c) 2023 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+
+import Mathlib.ModelTheory.Algebra.Field.CharP
+import Mathlib.FieldTheory.IsAlgClosed.AlgebraicClosure
+import Mathlib.FieldTheory.IsAlgClosed.Classification
+import Mathlib.ModelTheory.Satisfiability
+import Mathlib.Data.Nat.PrimeFin
+
+/-!
+
+# The First Order Theory of Algebraically Closed Fields
+
+This file defines the theory of algebraically closed fields of characteristic `p`, as well
+as proving completeness of the theory and the Lefschetz Principle.
+
+## Main definitions
+
+* `FirstOrder.Language.Theory.ACF p` : the theory of algebraically closed fields of characteristic
+`p` as a theory over the language of rings.
+* `FirstOrder.Field.ACF_isComplete` : the theory of algebraically closed fields of characteristic
+`p` is complete whenever `p` is prime or zero.
+* `FirstOrder.Field.ACF0_realize_iff_infinite_ACF_prime_realize` : the Lefschetz principle.
+
+## Implementation details
+
+To apply a theorem about the model theory of algebraically closed fields to a specific
+algebraically closed field `K` which does not have a `Language.ring.Structure` instance,
+you must introduce the local instance `compatibleRingOfRing K`. Theorems whose statement requires
+both a `Language.ring.Structure` instance and a `Field` instance will all be stated with the
+assumption `Field K`, `CharP K p`, `IsAlgClosed K` and `CompatibleRing K` and there are instances
+defined saying that these assumption imply `Theory.field.Model K` and `(Theory.ACF p).Model K`
+
+## References
+The first order theory of algebraically closed fields, along with the Lefschetz Principle and
+the Ax-Grothendieck Theorem were first formalized in Lean 3 by Joseph Hua
+[here](https://github.com/Jlh18/ModelTheoryInLean8) with the master's thesis
+[here](https://github.com/Jlh18/ModelTheory8Report)
+
+-/
+
+variable {K : Type*}
+
+namespace FirstOrder
+
+namespace Field
+
+open Ring FreeCommRing BigOperators Polynomial Language
+
+/-- A generic monic polynomial of degree `n` as an element of the
+free commutative ring in `n+1` variables, with a variable for each
+of the `n` non-leading coefficients of the polynomial and one variable (`Fin.last n`)
+for `X`.  -/
+def genericMonicPoly (n : ℕ) : FreeCommRing (Fin (n + 1)) :=
+  of (Fin.last _) ^ n + ∑ i : Fin n, of i.castSucc * of (Fin.last _) ^ (i : ℕ)
+
+/-- Monic polynomials of degree `n` over `K` are equivalent to functions `Fin n → K`  -/
+noncomputable def monicPolyEquivFin [CommRing K] [Nontrivial K] (n : ℕ) :
+    { p : K[X] // p.Monic ∧ p.natDegree = n } ≃ (Fin n → K) :=
```

Comment:
This has nothing to do with model theory, right? Can you move that to an earlier file?

Reply:
It should be split into an Equiv between the LHS and [Polynomial.degreeLT](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Polynomial/Basic.html#Polynomial.degreeLT) (by subtracting X^n; eraseLead may be useful, or not), and a [Basis](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Basis.html#Basis) of degreeLT indexed by `Fin n`.
Also, this holds for Semiring K.

Reply:
Actually there is already [Polynomial.degreeLTEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Polynomial/Basic.html#Polynomial.degreeLTEquiv) which is basically defeq to the Basis.